src/packet_analysis/protocol/gre/GRE.cc

```diff
@@ -150,29 +152,39 @@ bool GREAnalyzer::AnalyzePacket(size_t len, const uint8_t* data, Packet* packet)
 				}
 			}
 
-		else if ( proto_typ == 0x8200 )
+		else if ( ((proto_typ & 0x8200) == 0x8200 && (proto_typ & 0x0F) == 0) ||
+		          ((proto_typ & 0x8300) == 0x8300 && (proto_typ & 0x0F) == 0 &&
+		           (proto_typ <= 0x8370)) ||
+		          (proto_typ == 0x9000) )
 			{
-			// ARUBA. Following headers seem like they're always a 26-byte 802.11 QoS header, then
-			// an 8-byte LLC header, then IPv4. There's very little in the way of documentation
-			// for ARUBA's header format. This is all based on the one sample file we have that
-			// contains it.
-			if ( len > gre_len + 34 )
+			// ARUBA. In all of the sample data that we have, types 0x8200-0x8370 (with the bottom
+			// half of the second byte always zero) are always followed by an 802.11 QoS
+			// header. That means we can strip off the GRE header and forward directly to that
+			// analyzer. Type 0x9000 seems to just contain raw data, or at least it's nothing that
+			// Wireshark is able to parse. Try forwarding it along the default path for the GRE
+			// analyzer, but don't expect much from it.
+			if ( len <= gre_len )
 				{
-				gre_link_type = DLT_EN10MB;
-				erspan_len = 34;
-
-				// TODO: fix this, but it's gonna require quite a bit more surgery to the GRE
-				// analyzer to make it more independent from the IPTunnel analyzer.
-				// Setting gre_version to 1 here tricks the IPTunnel analyzer into treating the
-				// first header as IP instead of Ethernet which it does by default when
-				// gre_version is 0.
-				gre_version = 1;
-				proto = (data[gre_len + 34] & 0xF0) >> 4;
+				Weird("truncated_GRE", packet);
+				return false;
+				}
+
+			// TODO: it might make more sense to add an aruba analyzer here. it would duplicate a
+			// bit of the code later in this analyzer but it could handle everything more cleanly
+			// and not need a bunch of special cases in this code.
+			is_aruba = true;
+			gre_link_type = DLT_EN10MB;
+			gre_version = 1;
```

Comment:
Is this actually reasonable? The gre_version is parsed from the header, so we should not override it. For the added pcap, wireshark says gre_version=0. We have other places in the IPTunnel analyzer where testing for (gre_version != 0) means there's IP fields (that's probably also buggy because gre_version is initialized with -1).

I realize this construct might work out currently, but it's very surprising to change a parsed field to something else that's not actually in the packet.

Reply:
Removed.